sdk/queue/priority_queue.go

```diff
@@ -0,0 +1,206 @@
+// Package queue  provides Vault plugins with a Priority Queue. It can be used
+// as an in-memory list of queue.Item sorted by their priority, and offers
+// methods to find or remove items by their key. Internally it uses
+// container/heap; see Example Priority Queue:
+// https://golang.org/pkg/container/heap/#example__priorityQueue
+package queue
+
+import (
+	"container/heap"
+	"errors"
+	"fmt"
+	"sort"
+	"sync"
+
+	"github.com/mitchellh/copystructure"
+)
+
+// ErrEmpty is returned for queues with no items
+var ErrEmpty = errors.New("queue is empty")
+
+// ErrDuplicateItem is returned when the queue attmepts to push an item to a key that
+// already exists. The queue does not attempt to update, instead returns this
+// error. If an Item needs to be updated or replaced, pop the item first.
+var ErrDuplicateItem = errors.New("duplicate item")
+
+// ErrItemNotFound is a struct that implements the error interface
+var _ error = &ErrItemNotFound{}
+
+type ErrItemNotFound struct {
```

Comment:
Why create a new type, vs `newErrItemNotFound` just invoking & returning `fmt.Errorf()`?

Reply:
It's also missing a comment

Reply:
The new type allows users to check the error type, to distinguish from an error popping an item versus popping an item by key, which wasn't found. Something like this:

```go
i,err:=queue.PopByKey("key")
if err != nil {
  if err.(type) != queue.ErrItemNotFound { 
    return err
  } 
  // item not found, move on
}
```

Reply:
I updated the code to remove `ErrItemNotFound` and related things